### PR TITLE
feat(aws): add s3 awsv2

### PR DESCRIPTION
**Pull request by andydunstall**
Integrates `awsv2` (which uses the AWS SDK) into Dragonfly for S3 cloud snapshots (see https://github.com/romange/helio/pull/146).

`ServerFamily` will initialise the AWS SDK on startup (if configured to use S3), which loads configuration and credentials.

Each proactor shares the same S3 client to avoid loading credentials per proactor. The underlying HTTP client is thread safe and non-blocking, and the rest of the S3 client is thread safe so sharing the S3 client across proactors is ok (its mostly stateless so there won't be any contention except a RW lock to get credentials).

# Testing
## 200GB snapshot :heavy_check_mark:
Provisioning a `r6g.8xlarge` and adding 200GB of random data (~6GB per shard) (`redis-cli debug populate 200000000 test 1000 rand`), then uploading/downloading snapshots for a few hours with `for i in $(seq 100); do redis-cli save df snapshot; redis-cli debug load {bucket}/snapshots/snapshot-summary.dfs; done`.

(Note use a fixed snapshot name to avoid paying high S3 storage costs. S3 should be in the same region to avoid paying for the network upload)

Benchmarks:
* 200GB save: 20-30% CPU, 3:50 minutes
  * Note saving a snapshot to disk then uploading was 7 minutes so this is an improvement
* 200GB load: 20-25% CPU, 4:30 minutes

As described in https://github.com/romange/helio/pull/146, we could improve performance by removing the MD5 content header and payload signature, though its a tradeoff of performance and security, and depends on the SOC2 requirements.

## Retries :heavy_check_mark: 
To test the client correctly reconnects, running the upload/download loop shows above via a proxy (such as [toxiproxy](https://github.com/Shopify/toxiproxy)), then breaking the networking to check the client retries and recovers.

Configuring the proxy to drop the network for 1 second every 10 seconds, meaning existing connections are closed and reconnects fail to connect in that one second. Running with a 10GB snapshot which takes ~30s to load and ~30s to save, meaning each upload/download reconnects ~3 times.

Repeating 100 times works ok. Seeing the SDK correctly backoff and retry:
```
W20230925 08:59:45.336876 1018123 http_client.cc:203] aws: http client: failed to connect; host=localhost; error=system:111
W20230925 08:59:45.336966 1018123 logger.cc:79] aws: AWSClient: Request failed, now waiting 800 ms before attempting again.
```

The above 200GB snapshot test also showed occasional 5xx errors from S3 that are retried correctly:
```
W20230925 11:01:51.256095 20199 logger.cc:78] aws: AWSErrorMarshaller: Encountered AWSError 'InternalError': We encountered an internal error. Please try again.
W20230925 11:01:51.256206 20199 logger.cc:81] aws: AWSXmlClient: HTTP response code: 500
W20230925 11:01:51.256295 20199 logger.cc:78] aws: AWSClient: Request failed, now waiting 0 ms before attempting again.
```

## Credentials :heavy_check_mark: 
Test the different credential types
  - [x] Environment variables
  - [x] Local profile
    - [x] Override profile with `AWS_PROFILE`
  - [x] EC2 metadata

## Error Handling :heavy_check_mark: 
- [x] Invalid credentials: Logs `aws: s3 write file: failed to create multipart upload: InvalidAccessKeyId` and `SAVE` returns `Failed to open write file` (The error is not retryable so the SDK won't retry)
- [x] Bucket not found: Logs `aws: s3 write file: failed to create multipart upload: NoSuchBucket` and `SAVE` returns `Failed to open write file` (The error is not retryable so the SDK won't retry)
- [x] S3 unreachable: If S3 is unreachable (eg a firewall blocking access), the SDK will retry 5 times, which takes ~30s (with backoff) before it fails and logs a `Failed to open write file` as above

## Non-Blocking :heavy_check_mark:
To verify uploads don't block the thread, we can run a `redis-cli save` loop, and also run a lightweight benchmark measuring the maximum latency. We should not see the maximum latency increase significantly (which would indicate a thread may be blocked).

Running a `PING` loop and `GET foo` loop, then measuring the latency when running the 200GB upload loop above, seeing latency consistently under 10ms. There is overhead to uploading snapshots though there doesn't seem to be any threads being blocked for a long time.

## Regression Tests :heavy_check_mark:
Adds a regression test for S3 snapshots (https://github.com/dragonflydb/controlplane/issues/250)

As described in https://github.com/dragonflydb/controlplane/issues/250, this uses S3 directly which seems preferable to using some mocked S3 like MinIO, which isn't guaranteed to match S3 and is a pain to setup.

To avoid requiring credentials to run locally, this only runs if `DRAGONFLY_S3_BUCKET` is configured, then we can setup CI to include AWS credentials and an S3 bucket to access. That bucket can also have a configured retention to cleanup objects uploaded etc.

### CODE REVIEW

File: src/server/detail/snapshot_storage.cc [hunk 1]
```diff
@@ -312,19 +324,29 @@ io::Result<std::vector<std::string>, GenericError> AwsS3SnapshotStorage::LoadPat
...
-  ec = bucket.ListAllObjects(
-      prefix, [&](size_t sz, std::string_view name) { keys.push_back(std::string(name)); });
-  if (ec) {
-    return nonstd::make_unexpected(GenericError{ec, "Couldn't list objects in S3 bucket"});
-  }
+  do {
+    Aws::S3::Model::ListObjectsV2Request request;
+    request.SetBucket(std::string(bucket_name));
+    request.SetPrefix(std::string(prefix));
+    if (continuation_token != "") {
```
Comment (romange) [inline src/server/detail/snapshot_storage.cc:335]: `.empty()`

File: src/server/detail/snapshot_storage.cc [hunk 2]
```diff
@@ -312,19 +324,29 @@ io::Result<std::vector<std::string>, GenericError> AwsS3SnapshotStorage::LoadPat
...
+    if (outcome.IsSuccess()) {
+      continuation_token = outcome.GetResult().GetNextContinuationToken();
+      for (const auto& object : outcome.GetResult().GetContents()) {
+        keys.push_back(object.GetKey());
+      }
+    } else {
+      return nonstd::make_unexpected(GenericError{"Failed list objects in S3 bucket: " +
+                                                  outcome.GetError().GetExceptionName()});
+    }
+  } while (continuation_token != "");
```
Comment (romange) [inline src/server/detail/snapshot_storage.cc:349]: `continuation_token.empty()`

File: src/server/server_family.cc [hunk 3]
```diff
@@ -90,6 +89,17 @@ ABSL_FLAG(ReplicaOfFlag, replicaof, ReplicaOfFlag{},
...
 
+ABSL_FLAG(string, s3_endpoint, "", "endpoint for s3 snapshots, default uses aws regional endpoint");
+// Disable EC2 metadata by default, or if a users credentials are invalid the
+// AWS client will spent 30s trying to connect to inaccessable EC2 endpoints
+// to load the credentials.
+ABSL_FLAG(bool, s3_ec2_metadata, false,
+          "whether to load credentials and configuration from EC2 metadata");
+// Disables S3 payload signing (even over HTTP). This reduces the latency and
+// resource usage when writing snapshots to S3, at the expense of security.
+ABSL_FLAG(bool, s3_disable_payload_signing, false,
```
Comment (romange) [inline src/server/server_family.cc:100]: the name a bit confusing. maybe reverse the name to "s3_payload_signing" and flip the default to "true"?

File: src/server/detail/snapshot_storage.cc (hunk 2 above)
Comment (andydunstall) [reply src/server/detail/snapshot_storage.cc:349]: yep updated

File: src/server/detail/snapshot_storage.cc (hunk 1 above)
Comment (andydunstall) [reply src/server/detail/snapshot_storage.cc:335]: yep updated

File: src/server/server_family.cc (hunk 3 above)
Comment (andydunstall) [reply src/server/server_family.cc:100]: yep updated

**Review by romange — APPROVED** [review]

**Comment by andydunstall** [comment]
Doing some final testing found issues with CRC32 (which I'm fairly sure worked before but maybe was only testing against MinIO where it works ok): https://github.com/romange/helio/pull/152

**Review by romange — APPROVED** [review]

**Pull request merged**